iree/compiler/Conversion/Common/LinalgBufferizePass.cpp

```diff
@@ -49,458 +50,528 @@ namespace mlir {
+              ie = mappedTensors.member_end();
+         it != ie; ++it) {
+      tensors.push_back(getValue(*it));
     }
+    return tensors;
+  }
 
-    // If resultTensor already has a buffer, just use that.
-    Value alloc = bvm.lookupOrNull(resultTensor);
-    if (!alloc) {
-      Type resultType = en.value();
-      auto tensorType = resultType.dyn_cast<RankedTensorType>();
-      auto tensorShape = tensorType.getShape();
-      SmallVector<Value, 4> dynOperands;
-      for (auto dim : llvm::enumerate(tensorShape)) {
-        Value dimTensor = bvm.lookupOrNull(outTensor);
-        if (!dimTensor) dimTensor = outTensor;
-        if (dim.value() == TensorType::kDynamicSize) {
-          dynOperands.push_back(
-              b.createOrFold<memref::DimOp>(loc, dimTensor, dim.index()));
-        }
-      }
-      alloc = allocationFn(b, loc, tensorShape, tensorType.getElementType(),
-                           dynOperands);
-      bvm.map(resultTensor, alloc);
-    }
-    resultBuffers.push_back(alloc);
-
-    // Additionally, if the output buffer is used, clone its value for now.  The
-    // method `payloadUsesValueFromOutputOperandIndex` only works on named ops
-    // that have a region. Named ops like `conv`, etc. that are manually defined
-    // do not have this generated by default. So for now, just handled these
-    // manually defined ops specifically.
-    if (!isa<linalg::FillOp>(op.getOperation()) &&
-        op.payloadUsesValueFromOutputOperandIndex(resultIndex)) {
-      b.create<linalg::CopyOp>(loc, bvm.lookup(outTensor), alloc);
-    }
+  bool isEquivalent(Value v1, Value v2) {
+    return mappedTensors.isEquivalent(getPointer(v1), getPointer(v2));
   }
-  for (auto it : llvm::zip(op->getResults(), resultBuffers)) {
-    transferShapeOpsToMemref(b, std::get<0>(it), std::get<1>(it), bvm);
+
+  void insert(Value v) { mappedTensors.insert(getPointer(v)); }
+
+  void unionSets(Value v1, Value v2) {
+    mappedTensors.unionSets(getPointer(v1), getPointer(v2));
   }
-  return success();
-}
 
-// Non-conversion equivalent of the core MLIR Linalg bufferization patterns.
-static LogicalResult finalizeBufferAllocation(OpBuilder &b, linalg::LinalgOp op,
-                                              ValueRange inputs,
-                                              ValueRange outputs,
-                                              BlockAndValueMapping &bvm) {
-  SmallVector<Value, 8> newOperands = inputs;
-  newOperands.append(outputs.begin(), outputs.end());
-  auto otherOperands =
-      llvm::map_range(op.getAssumedNonShapedOperands(),
-                      [&bvm](Value v) { return bvm.lookupOrDefault(v); });
-  newOperands.append(otherOperands.begin(), otherOperands.end());
-  Location loc = op.getLoc();
-  op.clone(b, loc, {}, newOperands);
+  /// Sets the equivalance class that contains `v` as the set that contains the
+  /// result tensor of the dispatch region (i.e. a tensor that is the `value`
```

Comment:
I dont see why it is assuming that. It is clonging the operation. We dont have a check for this, but we should support multiple results. We might not have tested it, but Hanhan just filed an issue to address that. So this will be addressed soon.